xmtp_mls/src/identity/xmtp_id/identity.rs

```diff
@@ -71,112 +71,129 @@ impl Identity {
         self.signature_request.is_none()
     }
 
-    pub(crate) async fn create_to_be_signed<ApiClient: XmtpMlsClient + XmtpIdentityClient>(
+    /// Create a new [Identity] instance.
+    ///
+    /// If the address is not associated with an inbox_id, a new inbox_id will be generated.
+    /// Use legacy key to sign if it's present, otherwise users will be required to sign with their wallet.
+    ///
+    /// If the address is already associated with an inbox_id, the existing inbox_id will be used.
+    /// Users will be required to sign with their wallet.
+    ///
+    pub(crate) async fn new<ApiClient: XmtpMlsClient + XmtpIdentityClient>(
         address: String,
+        legacy_signed_private_key: Option<Vec<u8>>,
         api_client: &ApiClientWrapper<ApiClient>,
     ) -> Result<Self, IdentityError> {
-        // check if address is already associated with an inbox_id, generate a new inbox_id if not.
+        // check if address is already associated with an inbox_id
         let inbox_ids = api_client.get_inbox_ids(vec![address.clone()]).await?;
-        let mut is_new_inbox = false;
-        let mut nonce = 0;
-        let inbox_id;
-        if let Some(id) = inbox_ids.get(&address) {
-            inbox_id = id.clone();
-        } else {
-            is_new_inbox = true;
-            nonce = rand::random::<u64>();
-            inbox_id = generate_inbox_id(&address, &nonce);
-        }
-
+        let inbox_id = inbox_ids.get(&address);
         let signature_keys = SignatureKeyPair::new(CIPHERSUITE.signature_algorithm())?;
         let installation_public_key = signature_keys.public();
-        let member_identifier: MemberIdentifier = address.into();
-
-        let mut builder = SignatureRequestBuilder::new(inbox_id.clone());
-        if is_new_inbox {
-            builder = builder.create_inbox(member_identifier.clone(), nonce);
-        }
+        let member_identifier: MemberIdentifier = address.clone().into();
 
-        let mut signature_request = builder
-            .add_association(installation_public_key.to_vec().into(), member_identifier)
-            .build();
+        // if an inbox is not associated, we will create a new one
+        // and sign the installation key using legacy key or wallet.
+        if inbox_id.is_none() {
+            if let Some(legacy_signed_private_key) = legacy_signed_private_key {
+                // TODO: check if address matches legacy_signed_private_key
```

Comment:
is this TODO needed at all? Can we assume that `address` and `legacy_signed_private_key` passed in from platform SDK always match?

Reply:
They should always match in practice, but let's do the check just in case